core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -166,6 +168,29 @@ public Optional<Rel> visit(Cross cross) throws RuntimeException {
             .build());
   }
 
+  @Override
+  public Optional<Rel> visit(HashJoin hashJoin) throws RuntimeException {
+    var left = hashJoin.getLeft().accept(this);
+    var right = hashJoin.getRight().accept(this);
+    var leftKeys = hashJoin.getLeftKeys();
+    var rightKeys = hashJoin.getRightKeys();
+    var postFilter = hashJoin.getPostJoinFilter().flatMap(t -> visitExpression(t));
+    if (allEmpty(left, right, postFilter)) {
+      return Optional.empty();
+    }
+    return Optional.of(
+        ImmutableHashJoin.builder()
+            .from(hashJoin)
+            .left(left.orElse(hashJoin.getLeft()))
+            .right(right.orElse(hashJoin.getRight()))
+            .leftKeys(leftKeys)
+            .rightKeys(rightKeys)
+            .postJoinFilter(
+                Optional.ofNullable(
+                    postFilter.orElseGet(() -> hashJoin.getPostJoinFilter().orElse(null))))
```

Comment:
Because postFilter is already an Optional, we can use the `or` method instead of `Optional.ofNullable` here like:
```java
            .postJoinFilter(postFilter.or(hashJoin::getPostJoinFilter))
```

Reply:
You're not missing anything, I forget that we're on Java 8 sometimes so we don't get access to all the nice Optional util methods 🥲